### PR TITLE
Basic controller

### PR DESCRIPTION
**Pull request by engelmi**
This PR 
- adds a event loop (see [here](https://0pointer.net/blog/introducing-sd-event.html) for a good intro article)
- adds a basic controller accepting tcp connection requests 
- hookup the controller with the event loop
- updated readme

### CODE REVIEW

**Review by ArtiomDivak — APPROVED** [review]

File: src/common/dbus.h [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+#include "memory.h"
+
+#include <systemd/sd-bus.h>
+#include <systemd/sd-event.h>
+
+#define _cleanup_sd_event_ _cleanup_(sd_event_unrefp)
+#define _cleanup_sd_event_source_ _cleanup_(sd_event_source_unrefp)
+#define _cleanup_sd_bus_ _cleanup_(sd_bus_unrefp)
+#define _cleanup_sd_bus_slot_ _cleanup_(sd_bus_slot_unrefp)
+#define _cleanup_sd_bus_message_ _cleanup_(sd_bus_message_unrefp)
```
Comment (Darth-Mera) [inline src/common/dbus.h:13]: What is the purpose of the  underscores here?

Comment (engelmi) [reply src/common/dbus.h:13]: Using `_` for naming is called `snake case` - the [systemd codestyle](https://github.com/systemd/systemd/blob/main/docs/CODING_STYLE.md#code-organization-and-semantics) uses this (second bullet point).

Comment (Darth-Mera) [reply src/common/dbus.h:13]: Right, should've made my question more clear - snake case I understand, I specifically meant the prefix/suffix underscores. I've read some heated flame wars about using/not using `_foo_` naming

File: src/orch/controller.c [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+#include <errno.h>
+#include <netinet/in.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/socket.h>
+
+int create_master_socket(int port) {
+	struct sockaddr_in servaddr;
```
Comment (mkemel) [inline src/orch/controller.c]: Is it formatted with tabs?
Looking at the systemd coding style - they don't like tabs :)

File: src/common/dbus.h (hunk 1 above)
Comment (engelmi) [reply src/common/dbus.h:13]: Ah ok. That I can't definitely answer as I basically copy+pasted it from the POC. The [ISO/IEC 9899:201x](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1570.pdf) (section 7.1.3) could give an explanation.

File: src/orch/controller.c (hunk 2 above)
Comment (engelmi) [reply src/orch/controller.c]: Good catch! Seems like I've copied the wrong `.clang-format` file previously. Updated.

File: src/common/dbus.h (hunk 1 above)
Comment (sdunnagan) [reply src/common/dbus.h:13]: There are similar macros defined in systemd.  For example, in coccinelle/macros.h:
`#define _cleanup_(x) __attribute__((__cleanup__(x)))`

Comment (Darth-Mera) [reply src/common/dbus.h:13]: Ah, got it, thanks, @sdunnagan !

**Review by Darth-Mera — APPROVED** [review]

File: src/orch/controller.c [hunk 3]
```diff
@@ -0,0 +1,83 @@
+#include "controller.h"
+
+#include <errno.h>
+#include <netinet/in.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/socket.h>
+
+int create_master_socket(int port) {
```
Comment (sdunnagan) [inline src/orch/controller.c]: The port number isn't changed by this function, so this variable should be made const:
`int create_master_socket(const int port)`

File: src/orch/orch.c [hunk 4]
```diff
@@ -1,7 +1,47 @@
...
 #include <stdio.h>
 #include <stdlib.h>
 
-int main() {
+int getArgs(int argc, char *argv[], int *ctrlPort) {
+	if (argc < 2) {
+		fprintf(stderr, "No controller port given\n");
+		return -1;
+	}
+	*ctrlPort = atoi(argv[1]);
```
Comment (sdunnagan) [inline src/orch/orch.c]: As mentioned in the systemd coding standards, we shouldn't use plain atoi().  Instead, we can define a safe version of atoi like systemd does, in src/basic/parse-util.c:
`int safe_atoi(const char *s, int *ret_i)`

File: src/orch/orch.c [hunk 5]
```diff
@@ -1,7 +1,47 @@
+#include "../common/dbus.h"
+#include "controller.h"
+
 #include <stdio.h>
 #include <stdlib.h>
 
-int main() {
+int getArgs(int argc, char *argv[], int *ctrlPort) {
```
Comment (sdunnagan) [inline src/orch/orch.c]: Maybe use snake_case here for consistency with systemd: get_args().  Also, consider using getopt().

File: src/orch/controller.c (hunk 3 above)
Comment (engelmi) [reply src/orch/controller.c]: Changed.

File: src/orch/orch.c (hunk 4 above)
Comment (engelmi) [reply src/orch/orch.c]: Good point. I tried to implement a safer version via [strtol](https://man7.org/linux/man-pages/man3/strtol.3.html), which seemed pretty straight forward and allows better error detection. Not sure about disadvantages. Please have a look at `src/common/parse-util.h`

File: src/orch/orch.c (hunk 5 above)
Comment (engelmi) [reply src/orch/orch.c]: Changed the naming and tried to implement it via `getopt()`. Moved it to a dedicated file. If this is ok, I'd try to implement basic tests for it in a follow-up PR.

**Pull request merged**